### DISCUSSION TITLE
Failed to send email using Postmark

### DISCUSSION THREAD

**Discussion by kkostov** [umputun/remark42#1359] [Q&A]
I'm having trouble configuring remark to use Postmark's [SMTP](https://postmarkapp.com/developer/user-guide/send-email-with-smtp). The operation fails with a tls related error:

```
2022/05/15 08:36:22.369 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to make smtp client: failed to dial smtp tls to smtp.postmarkapp.com:587: tls: first record does not look like a TLS handshake - 500 - xxx.xxx.xxx.xx - xxx [caused by auth/provider/verify.go:188 provider.VerifyHandler.sendConfirmation]
```

My SMTP settings are as follows:

```
SMTP_HOST=smtp.postmarkapp.com
SMTP_PORT=587
SMTP_TLS=true
SMTP_USERNAME=***
SMTP_PASSWORD=***
```

I did find a few other issues discussing TLS but I couldn't find a way to make it work.

Postmark offers [two main ways](https://postmarkapp.com/support/article/1207-how-to-create-and-send-through-message-streams#sendingsmtp) to authenticate, the one that seems compatible with remark is "SMTP Token".

I tried using any of the supported ports - 25, 2525 and 587 as well was setting both SMTP_TLS=true and/or SMTP_STARTTLS=true.

Is it possible that I'm overlooking something? 

Thanks

**Comment by paskal** [comment]
Port 587 is used with StartTLS, 465 is the standard TLS port. Please try using 465.

**Comment by kkostov** [comment]
I did try that as well (it was a recommendation in other issues), however port 465 [doesn't seem to be in use by Postmark](https://postmarkapp.com/blog/choosing-the-right-smtp-port#smtp-port-465-a-deprecated-port-you-should-no-longer-use).

> Don’t use port 465 for SMTP if you don’t have to. Many modern hosting providers and email service providers block it or don’t support it (including Postmark).

Calls using 465 result in a timeout (as if there is no service listening there).

The [guideline](https://postmarkapp.com/blog/choosing-the-right-smtp-port#smtp-port-587-the-right-choice-for-nearly-every-situation) from the postmark docs appears to be to use 587:

> Port 587 was established as a modern secure SMTP port for message delivery. It supports TLS natively and STARTTLS as well, allowing for the secure submission of mail over SMTP.

**Comment by umputun** [comment]
well, apparently this "supports TLS natively and STARTTLS as well" doesn't work with [go-pkgz/email](https://github.com/go-pkgz/email) library which is a thin wrapper on top of go stdlib. If you can reproduce the issue with some small snippet you can submit a ticket to that library.

Another thing - what version of remark42 do you use? Support of STARTTLS is not in latest, but in master only.

_moved to discussions_

**Reply by paskal** [reply]
Indeed, with the `master` tag of Docker image `SMTP_STARTTLS=true` should work.

**Reply by kkostov** [reply]
On master, using 
```
- SMTP_HOST=smtp.postmarkapp.com
- SMTP_PORT=587/2525/25
- SMTP_STARTTLS=true
```

The following error is raised when sending an email:

```
2022/05/18 05:33:49.655 [WARN]  {sender/email.go:76 sender.(*Email).Send.func1} can't close smtp connection, close tcp XXX.X.X.X:60352->34.249.162.51:587: use of closed network connection
2022/05/18 05:33:49.655 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to auth to smtp smtp.postmarkapp.com:587: unencrypted connection - 500 - XXX.X.X.X - /auth/email/login?site=blog&address=xx@xxxx.xx&user=me [caused by auth/provider/verify.go:188 provider.VerifyHandler.sendConfirmation]
```

I was also able to reproduce the behaviour using the integration test `TestSendIntegration` for sending emails in the project. I will try to create a repro using only [go-pkgz/email](https://github.com/go-pkgz/email) and report the issue as suggested.

In the meantime, do you see a possible work-around?
My current thinking is to fork remark42 in order to substitute the email `Sender` with something like [this](https://github.com/ainsleyclark/go-mail). Would you be interested in a PR with this or is it out of scope for the project?

**Reply by paskal** [reply]
I found the problem, I'll see if it's specific to this provider or if the library is off and will try to fix it. We store the hostname of the server as "host:port", and then on TLS check, it's failing because the host itself announces only "host" about itself.

**Reply by paskal** [reply]
Once https://github.com/go-pkgz/email/pull/12 is merged I'll bump the library here and `master` will start working. That's the downside of using `master`: you're testing new features, but there is a chance that you're the first person actually using them so thanks to you we found a generic problem with StartTLS which became enabled just three weeks ago.

**Reply by paskal** [reply]
The new version with the fix is built and pushed. Please pull the latest `master` and let us know if there are any problems.

**Reply by kkostov** [reply]
Hi, the problem still persists for me.

I'm not sure if both SMTP_TLS and STARTLS should set be set to true so I experimented with both combinations on port 2525 and 587 but without success.

```
2022/05/23 07:09:48.358 [WARN]  {sender/email.go:76 sender.(*Email).Send.func1} can't close smtp connection, close tcp 172.19.0.2:46474->63.35.51.189:2525: use of closed network connection
2022/05/23 07:09:48.358 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to auth to smtp smtp.postmarkapp.com:2525: unencrypted connection - 500 - <SERVER IP>
```

```
2022/05/23 07:09:20.526 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to make smtp client: failed to dial smtp tls to smtp.postmarkapp.com:2525: tls: first record does not look like a TLS handshake - 500 - <SERVER IP> - /auth/email/login?site=blog&address=XXXX@XXXX&user=xxx [caused by auth/provider/verify.go:188 provider.VerifyHandler.sendConfirmation]
2
```

```
2022/05/23 07:04:13.002 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to make smtp client: failed to dial smtp tls to smtp.postmarkapp.com:587: tls: first record does not look like a TLS handshake - 500 - <SERVER IP> - /auth/email/login?site=blog&address=XXXX@XXXX&user=xxx [caused by auth/provider/verify.go:188 provider.VerifyHandler.sendConfirmation]
2
```

```
2022/05/23 07:03:21.743 [WARN]  {sender/email.go:76 sender.(*Email).Send.func1} can't close smtp connection, close tcp 172.19.0.2:47346->34.249.162.51:587: use of closed network connection
2022/05/23 07:03:21.743 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to auth to smtp smtp.postmarkapp.com:587: unencrypted connection - 500 - 162.158.233.73 - /auth/email/login?site=blog&address=XXXX@XXXX&user=xxx [caused by auth/provider/verify.go:188 provider.VerifyHandler.sendConfirmation]
2
```

**Reply by paskal** [reply]
SMTP_STARTTLS should be true and port should be 587, that's it.

**Reply by paskal** [reply]
To be precise, that is what I have in my settings:
```yaml
      - NOTIFY_ADMINS=email
      - ADMIN_SHARED_EMAIL=me@example.org
      - SMTP_HOST=smtp.postmarkapp.com
      - SMTP_PORT=587
      - SMTP_STARTTLS=true
      - SMTP_USERNAME=PM-T-outbound-*****--**********
      - SMTP_PASSWORD=******
      - NOTIFY_EMAIL_FROM=sender@example.org
```
This is the log entry:
```
remark42-dev | 2022/05/23 14:24:03.579 [DEBUG] {notify/email.go:159 notify.(*Email).buildAndSendMessage} send notification via email: with username 'PM-T-outbound-*****--**********' at server smtp.postmarkapp.com:587 with StartTLS, comment id 792871bd-4e3e-4c1c-ac08-8ebd79bcb2f7
```
This is the Postmark log:
<img width="1319" alt="image" src="https://user-images.githubusercontent.com/712534/169892486-ee90bd77-9642-46bd-b262-7e1ab1e9322e.png">
This is the resulting email:
<img width="631" alt="image" src="https://user-images.githubusercontent.com/712534/169892435-ebeda4ae-389e-46ec-8d7a-90615d0378df.png">
I'm using latest `master` image, here is a line from log:
```
remark42-dev | remark42 master-3b5f44d-20220520T16:02:32
```

**Reply by kkostov** [reply]
Thanks for the detailed example!
Indeed that's exactly the config I tried, however the confirmation email when signing up for an account using email is not succeeding:

```
2022/05/25 02:06:38.710 [WARN]  {sender/email.go:76 sender.(*Email).Send.func1} can't close smtp connection, close tcp 172.19.0.2:35588->34.249.162.51:587: use of closed network connection
2022/05/25 02:06:38.710 [INFO]  {rest/httperrors.go:39 rest.SendErrorJSON} failed to send confirmation - failed to auth to smtp smtp.postmarkapp.com:587: unencrypted connection - 500 - xxx.xxx.xxx.xx - /auth/email/login?site=blog&address=XXXXXX&user=XXXXX [caused by auth/provider/verify.go:188 provider.VerifyHandler.sendConfirmation]
```

**Reply by paskal** [reply]
Please verify that you did a docker pull for the master image and using master and not latest. Unless the version in the log on the start is the same as the one I posted in the last line, it won't work.

**Reply by kkostov** [reply]
Indeed, the version hash matches the one from your example and the last commit on master:

```
changing ownership of /srv/var to app:app (remark42 user inside the container)
execute "/srv/remark42 server"
remark42 master-3b5f44d-20220520T16:02:30
2022/05/25 02:01:21.356 [INFO]  {cmd/server.go:289 cmd.(*ServerCommand).Execute} start server on port :8080

```

**Reply by paskal** [reply]
Then something is off about remark42 configuration, please post complete docke-compose file you're using, with hiding the passwords and secret values in it.

**Reply by kkostov** [reply]
This is the docker compose file I'm using:

```yaml
version: "2"

services:
  remark:
    image: umputun/remark42:master
    container_name: "remark42"
    restart: always
    ports:
      - "8181:8080"

    environment:
      - REMARK_URL=https://xxxxxxx
      - SITE=blog,remark
      - SECRET=xxxxxxx
      - STORE_BOLT_PATH=/srv/var/db
      - BACKUP_PATH=/srv/var/backup
      - DEBUG=true
      - AUTH_GITHUB_CID=xxxxxxx
      - AUTH_GITHUB_CSEC=xxxxxxx
      - ADMIN_SHARED_ID=xxxxxxx
      - AUTH_TWITTER_CID=xxxxxxx
      - AUTH_TWITTER_CSEC=xxxxxxx
      - ADMIN_SHARED_EMAIL=xxxxxxx@xxxxxxx.xx
      - NOTIFY_ADMINS=email
      - AUTH_EMAIL_FROM=xxxxxxx@xxxxxxx.xx
      - AUTH_EMAIL_ENABLE=true
      - NOTIFY_EMAIL_FROM=xxxxxxx@xxxxxxx.xx
      - SMTP_HOST=smtp.postmarkapp.com
      - SMTP_PORT=587
      - SMTP_STARTTLS=true
      - SMTP_USERNAME=xxxxxxx
      - SMTP_PASSWORD=xxxxxxx
    volumes:
      - ./var:/srv/var
```

**Reply by paskal** [reply]
This configuration **must** work with the current version of remark42 and couldn't produce the `unencrypted connection` error you posted, it's not possible. I propose you try running dev installation from scratch in a different directory and try to make email work there with just the variables I provided, and then building up configuration till it looks same as your production so you'll know where the problem is introduced. I assume even with the very same configuration new docker-compose would work, as the problem is with something else, like some environment variable not propagated to remark42 binary for some reason.

**Reply by kkostov** [reply]
I share the feeling - there must be something else that’s not right. I will have a closer look and try to debug it. In any case, thanks a lot for the guidance and examples above, I really appreciate it 🙌!
